rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,144 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                    // Skip already taken from log.
-                    if visited_ids.contains(&merged_id) {
-                        continue;
-                    }
+                &merged_oids_holder
+            }
+        }
+        .iter()
+        .filter(|oid| !deleted_offset_ids.contains(oid))
+        // Truncate the offset ids using offset and limit.
+        // The .skip(...) and .take(...) will return an empty iterator if the offset and limit is out of bound.
+        .skip(input.offset.unwrap_or(u32::MIN) as usize)
+        .take(input.limit.unwrap_or(u32::MAX) as usize);
 
-                    let user_id = match record_segment_reader
-                        .get_user_id_for_offset_id(merged_id)
-                        .await
-                    {
-                        Ok(user_id) => user_id,
-                        Err(e) => {
-                            println!("Error reading Record Segment: {:?}", e);
-                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
-                        }
-                    };
-                    ids.push(user_id.to_string());
+        // Hydrate data
+        // This should be a mapping from offset id to the index of the document in the final result.
+        let truncated_offset_id_order: HashMap<u32, usize> = merged_offset_ids
+            .enumerate()
+            .map(|(i, offset_id)| (*offset_id, i))
+            .collect();
+        let mut ids: Vec<String> = vec![String::new(); truncated_offset_id_order.len()];
+        let mut metadata = Vec::new();
+        let mut documents = Vec::new();
+        let mut logged_offset_ids: HashSet<u32> = HashSet::new();
+        if input.include_metadata {
+            metadata = vec![None; truncated_offset_id_order.len()];
+            documents = vec![None; truncated_offset_id_order.len()];
+        }
 
-                    if input.include_metadata {
-                        let record = match record_segment_reader
-                            .get_data_for_offset_id(merged_id)
-                            .await
-                        {
-                            Ok(record) => record,
-                            Err(e) => {
-                                tracing::error!("Error reading Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentReadError,
-                                );
-                            }
-                        };
-                        metadata.push(record.metadata.clone());
-                        match record.document {
-                            Some(document) => documents.push(Some(document.to_string())),
-                            None => documents.push(None),
-                        }
-                    }
+        // Hydrate the data from the materialized logs first.
+        for (log, _) in mat_records.iter() {
```

Comment:
I think this could be made more efficient? Say you have 10k entries on the log and If we are only limit 100, than this does not need to traverse over the whole log again right?